sdk/keyvault/azure-keyvault-keys/tests/test_keys_async.py

```diff
@@ -258,9 +263,14 @@ async def test_backup_restore(self, azure_keyvault_url, **kwargs):
         restored_key = await self._poll_until_no_exception(restore_function, expected_exception=ResourceExistsError)
         self._assert_key_attributes_equal(created_bundle.properties, restored_key.properties)
 
+    @parameterized.expand([param(is_hsm=b) for b in [True, False]], name_func=suffixed_test_name)
     @KeyVaultPreparer()
     async def test_key_list(self, azure_keyvault_url, **kwargs):
-        client = self.create_client(azure_keyvault_url)
+        is_hsm = kwargs.pop("is_hsm")
+        self._skip_if_not_configured(is_hsm)
+        azure_keyvault_url = self.managed_hsm_url if is_hsm else azure_keyvault_url
```

Comment:
This method demands an argument but throws it away half the time because it knows better than the caller what its value should be. Could the caller instead determine the correct value?

Reply:
The `azure_keyvault_url` parameter is automatically passed in as a positional parameter by the PowerShellPreparer, and the name matches the vault URL environment variable. I tried a couple different ways of fetching the MHSM URL through the preparer and passing that in as well, but wasn't able to get it working unfortunately

Reply:
Perhaps the (another?) problem then is that PowerShellPreparer requires the test to accept an argument it doesn't need? I don't have a solution in mind, I'm just poking at this confusion around who's responsible for figuring out client parameters👉 👉

Reply:
Yeah, it's not the best setup. I'd definitely like to dig into this when possible, but for now I don't see it as a blocker per se

Reply:
Getting this right is worth the effort. I think that's appropriate effort for this PR because it's the first to confront this inflexibility of the current code. If you kick the can today, you can expect that someone else, probably future you, will have to choose between taking the extra time to refactor this kludge out of existence or adding another quick kludge on top of it. And in the meantime this kludge is just hanging out looking awkward :flushed:

Reply:
You're making points, and I know my future self appreciates it 😞

Reply:
After brainstorming and messing with the preparer, I think the best approach here might actually be to not use the PowerShellPreparer for these tests. The main purpose of it is to unify/simplify fetching environment variables, but we already have to subvert that by manually getting the HSM URL and registering it with the scrubber -- which we have to do, because the PowerShellPreparer can't handle getting environment variables for two different directories (Key Vault and Managed HSM) simultaneously. 

Avoiding the preparer for these tests would unify the variable fetching approach, and it would also leave endpoint setting entirely to the test method. I think that makes sense, since the method is the only actor that's aware of whether or not it should target an HSM (via `is_hsm`). This would require re-implementing the preparer's [`_set_mgmt_settings_real_values` method](https://github.com/Azure/azure-sdk-for-python/blob/master/tools/azure-sdk-tools/devtools_testutils/powershell_preparer.py#L52-L62), but that could be done in the new KeysTestCase class.